cumulus/client/consensus/aura/src/collators/mod.rs

```diff
@@ -51,29 +53,34 @@ async fn check_validation_code_or_log(
 				%para_id,
 				"Failed to fetch validation code hash",
 			);
-			return
+			return None
 		},
 	};
 
 	match state_validation_code_hash {
-		Some(state) =>
+		Some(state) => {
 			if state != *local_validation_code_hash {
 				tracing::warn!(
 					target: super::LOG_TARGET,
 					%para_id,
 					?relay_parent,
 					?local_validation_code_hash,
 					relay_validation_code_hash = ?state,
-					"Parachain code doesn't match validation code stored in the relay chain state",
+					"Parachain code doesn't match validation code stored in the relay chain state. This is expected if you are using an override for example.",
```

Comment:
Is it expected actually? If we look at how `lookahead` collator is typically instantiated, we pass `client.code_at` as code_hash_provider like here:
https://github.com/paritytech/polkadot-sdk/blob/b9c81e1fafd2e03a1d452302e595d4e470c0d9fc/cumulus/polkadot-parachain/src/service.rs#L830-L832

With your changes in `client.rs` this will already take substitutes and overrides into account. The hash should match then and we would not even need to return the code hash from the relay chain, right?

Reply:
If you are using a `override`, you will never find the hash of the override in the relay chain state. However, the relay chain would reject your pov if you send it with the hash of you override, as it doesn't know this hash.

Reply:
So what is not super clear to me is what is the point of this check here except the logging message.

We fetch the code hash from the relay chain here and then submit it to the relay chain along with the collation. Previously if there is some other reason for a code hash mismatch, it would be caught. Can we not check if there is an override or substitute in place and only in that case return the relay chain code hash here?

Reply:
> So what is not super clear to me is what is the point of this check here except the logging message.

The entire purpose of this function was always to just log. There is nothing we can do, besides informing the human that something is wrong.

> Previously if there is some other reason for a code hash mismatch, it would be caught.

Not sure what other reason you mean, but we would still catch it. I mean even if the parachain is using a completely different code in its state, it will be printed. There is no real difference to before. The only difference is that we tell the relay chain what it expects from us (a little bit like cheating). However, the worst case would be that the relay chain rejects the Pov.